pkg/dns/aws/dns.go

```diff
@@ -0,0 +1,248 @@
+	ELB     *elb.ELB
+	Route53 *route53.Route53
+
+	config Config
+	lock   sync.RWMutex
+	// zones is a cache of all known DNS zones and their associated tags.
+	zones []*zoneWithTags
+	// updatedRecords is a cache of records which have been created during the
+	// life of this manager. The key is zoneID+domain+target. This is a quick hack
+	// to minimize AWS API calls for now.
+	updatedRecords map[string]struct{}
+}
+
+// Config is the necessary input to configure the manager.
+type Config struct {
+	// AccessID is an AWS credential.
+	AccessID string
+	// AccessKey is an AWS credential.
+	AccessKey string
+	// Region is an AWS config.
+	Region string
+
+	// BaseDomain should be the *absolute* name shared by the zones.
+	// Example: 'devcluster.openshift.com.'
+	BaseDomain string
+	// ClusterID is the value of the 'tectonicClusterID' tag identifying the
+	// cluster.
+	ClusterID string
+}
+
+func NewManager(config Config) (*Manager, error) {
+	creds := credentials.NewStaticCredentials(config.AccessID, config.AccessKey, "")
+	sess, err := session.NewSession(&aws.Config{
+		Credentials: creds,
+		Region:      aws.String(config.Region),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create AWS client session: %v", err)
+	}
+	return &Manager{
+		ELB:     elb.New(sess),
+		Route53: route53.New(sess),
+
+		config:         config,
+		zones:          []*zoneWithTags{},
+		updatedRecords: map[string]struct{}{},
+	}, nil
+}
+
+type zoneWithTags struct {
+	Zone *route53.HostedZone
+	Tags []*route53.Tag
+}
+
+// updateZoneCache caches all zones and their tags.
+func (m *Manager) updateZoneCache() error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// There's no reason to update this more than once as we're only interested in
+	// the zones scoped to this cluster.
+	if len(m.zones) > 0 {
+		return nil
+	}
+
+	// Find and cache all zones and their tags.
+	// TODO: probably only need to store the two zones we care about.
+	zones := []*zoneWithTags{}
+	listedAllTags := true
+	f := func(resp *route53.ListHostedZonesOutput, lastPage bool) (shouldContinue bool) {
+		for _, zone := range resp.HostedZones {
+			tagsList, err := m.Route53.ListTagsForResource(&route53.ListTagsForResourceInput{
+				ResourceType: aws.String("hostedzone"),
+				ResourceId:   zone.Id,
+			})
+			if err != nil {
+				logrus.Errorf("failed to list tags for hosted zone %s: %v", zone.Id, err)
+				listedAllTags = false
+				return false
+			}
+			zones = append(zones, &zoneWithTags{Zone: zone, Tags: tagsList.ResourceTagSet.Tags})
+		}
+		return true
+	}
+	if !listedAllTags {
+		return fmt.Errorf("failed to list all hosted zone tags")
+	}
+	err := m.Route53.ListHostedZonesPages(&route53.ListHostedZonesInput{}, f)
+	if err != nil {
+		return fmt.Errorf("failed to list hosted zones: %v", err)
+	}
+
+	m.zones = zones
+	return nil
+}
+
+// findClusterPublicZone finds the public zone given the base domain.
+func (m *Manager) findClusterPublicZone() *route53.HostedZone {
```

Comment:
I was hoping to send something like this upstream to external-dns ... add the ability to lookup zones by tags. In the meantime, the code will live in the operator.